test/distributed/_composable/fsdp/test_fully_shard_comm.py

```diff
@@ -655,6 +656,128 @@ def test_fully_shard_multi_module_backward_prefetch(self):
                 optim.step()
                 optim.zero_grad()
 
+    @skip_if_lt_x_gpu(2)
+    def test_fully_shard_with_forward_prefetch(self):
+        n_layers = 3
+        reshard_after_forward = True
+        checkpoint_impl = "utils"
+        model, optim, inp = self._init_transformer(
+            n_layers, reshard_after_forward, checkpoint_impl
+        )
+
+        # Show a lightweight implementation of forward prefetching relying on
+        # two data structures and a bool tracking the first iteration
+        pre_forward_order: List[FSDPModule] = []
+        module_to_pre_forward_index: Dict[FSDPModule, int] = {}
+        is_first_iter = True
+
+        def forward_pre_hook(module: nn.Module, args):
+            if is_first_iter and module not in module_to_pre_forward_index:
+                module_to_pre_forward_index[module] = len(pre_forward_order)
+                pre_forward_order.append(module)
+            elif not is_first_iter:
+                pre_forward_index = module_to_pre_forward_index.get(module, -1)
+                if pre_forward_index < 0:
+                    return
+                target_pre_forward_index = pre_forward_index + 1
+                if target_pre_forward_index >= len(pre_forward_order):
+                    return
+                target_module = pre_forward_order[target_pre_forward_index]
+                assert isinstance(target_module, FSDPModule), f"{target_module}"
+                # TODO: We need to access the training state to avoid
+                # prefetching in backward when composing with AC
+                if fully_shard.state(target_module)._training_state in (
+                    TrainingState.IDLE,
+                    TrainingState.FORWARD,
+                ):
```

Comment:
This is the only internal state that we need to give to users to be able to implement forward prefetching on top of FSDP2.